Allow the translation of enum value used for elementLabelProp in ExpandPanelRenderer

Maxouwell opened the pull request:
PR for #2288

sdirix commented on packages/material-renderers/src/layouts/ExpandPanelRenderer.tsx:
```diff
@@ -296,12 +299,35 @@ export const withContextToExpandPanelProps = (
+      );
+      if (
+        childSchema &&
+        childSchema.type === 'string' &&
```
The check for the type is not needed

Maxouwell replied:
Fixed

sdirix commented on packages/material-renderers/src/layouts/ExpandPanelRenderer.tsx:
```diff
@@ -296,12 +299,35 @@ export const withContextToExpandPanelProps = (
-      : get(childData, getFirstPrimitiveProp(schema), '');
+
+    let childLabel;
+    if (childLabelProp) {
```
The label determination should be within a `useMemo`

Maxouwell replied:
Fixed

sdirix commented on packages/material-renderers/src/layouts/ExpandPanelRenderer.tsx:
```diff
@@ -296,12 +299,35 @@ export const withContextToExpandPanelProps = (
+      if (
+        childSchema &&
+        childSchema.type === 'string' &&
+        (childSchema.enum !== undefined || childSchema.const !== undefined)
```
Currently we only check for the `enum` case but we should also cover the "one of enums" which are pairs of const an optional title properties, similar to the testers of the enum and oneofenum controls.

Then we need to call the `enumToEnumOptionMapper` or the `oneOfToEnumOptionMapper` accordingly.

Maxouwell replied:
I don't have case of enumOf on my side, I'll look to find some in the examples, and add it.

Maxouwell replied:
Added the enumOf case.

Note: the default behavior when no translating is setup will change (showing `const` value => showing `title` value instead). I think it's better but your mileage may vary.

sdirix replied:
Its definitely better ;)

sdirix reviewed (commented):
Looks good already. However we should also support the one of enum case

Maxouwell commented:
Resolved the conflicts, rebased and squasshed

sdirix reviewed (changes requested):
I had a look, improved the types and aligned the function declarations to arrow style as the rest of the code base.

I also noticed the following behavior.

Assume the following JSON Schema in use:
```js
{
  type: 'object',
  properties: {
    comments: {
      type: 'array',
      title: 'Messages',
      items: {
        type: 'object',
        properties: {
          message1: {
            type: 'string',
          },
          message2: {
            type: 'string',
            enum: ['foo', 'bar'],
          },
        },
      },
    },
  },
}
```

Then the auto generated keys for the enum will be `comments.message2.foo` and `comments.message2.bar`.

However for the summary title in the expand panel, the following keys will be used: `comments.foo` and `comments.bar`. Therefore if a proper translation of the enum is used, it will not automatically show up in the expand panel, instead another key needs to be additionally translated.

I assume that this was not on purpose?

Can you fix the issue? Ideally the test case should also be improved. In the moment there is just a mock translator in there which does not check the handed over keys. I assume that is why the issue was overlooked.

Maxouwell commented:
You were right, the previous code had problems on translation key, and on elementLabelProp on deeper path.

I fixed it and added some tests for the deep cases.

Maxouwell commented on packages/material-renderers/test/renderers/MaterialArrayLayout.test.tsx:
```diff
@@ -738,4 +910,236 @@ describe('Material array layout', () => {
+    expect(getChildLabel(wrapper, 0)).toBe(
+      'translator.root.messageType.MSG_TYPE_1'
+    );
+    // TODO Possible impact to existing jsonForm setup, previously if the translate fonction was setup,
+    //  the enumOf title was written as-is and not pushed to the translation function with a key.
+    //  The oneOf title is pushed for translation over the const value in renderer.ts/oneOfToEnumOptionMapper(),
+    //  seems a bug to me
+    expect(getChildLabel(wrapper, 1)).toBe(
+      'translator.root.messageType.Second message type'
+    );
```
I came across this while fixing the code, I think it needs your opinion.

sdirix replied:
The title is preferred over the value as in practice the value is often a non-string like an object while the title will always be a string. If that is not wanted, the user can define an `i18n` key in the schema.

sdirix commented on packages/material-renderers/test/renderers/MaterialArrayLayout.test.tsx:
```diff
@@ -738,4 +910,236 @@ describe('Material array layout', () => {
+    expect(getChildLabel(wrapper, 0)).toBe(
+      'translator.root.messageType.MSG_TYPE_1'
+    );
+    // TODO Possible impact to existing jsonForm setup, previously if the translate fonction was setup,
+    //  the enumOf title was written as-is and not pushed to the translation function with a key.
+    //  The oneOf title is pushed for translation over the const value in renderer.ts/oneOfToEnumOptionMapper(),
+    //  seems a bug to me
```
```suggestion
```

sdirix reviewed (commented):
I found one minor and one major issue with the current state. Besides that it looks good!

Maxouwell commented on packages/material-renderers/src/layouts/ExpandPanelRenderer.tsx:
```diff
@@ -334,6 +379,69 @@ export const withContextToExpandPanelProps = (
+) => {
+  const childData = Resolve.data(data, childPath);
+
+  if (!childLabelProp) return get(childData, getFirstPrimitiveProp(schema), '');
```
Fixed by https://github.com/eclipsesource/jsonforms/pull/2289/commits/27ad9db083068fe212d765e1b3a07dca750800f8

Maxouwell commented on packages/material-renderers/src/layouts/ExpandPanelRenderer.tsx:
```diff
@@ -334,6 +379,69 @@ export const withContextToExpandPanelProps = (
+  const fallbackI18nKey =
+    getI18nKeyPrefix(schema, uiSchema, childPath) +
+    '.' +
+    getI18nKeyPrefix(schema, uiSchema, childLabelProp);
```
Fixed by:
* https://github.com/eclipsesource/jsonforms/pull/2289/commits/72229bfaae733bd993810a8f441039fc6b02d103 for i18n on schema
* https://github.com/eclipsesource/jsonforms/pull/2289/commits/3091b2726333eae6b6e93566cf20ae09550c44e5 for i18n on uiSchema for enum (and test refactoring)

I didn't add the i18n on UiSchema for oneOf as I'm not sure it is coherent

Maxouwell commented:
I just saw that the same behaviour exist in ListWithDetail, and that the solution don't support if `elementLabelProp` point to an enum array (or oneOf array if it's a thing).
I propose we finish this PR and I will open others PRs for these other ones.

sdirix commented:
Update: I will re-review this PR again this week

sdirix commented on packages/core/src/util/label.ts:
```diff
@@ -81,3 +100,69 @@ const labelDescription = (text: string, show: boolean): LabelDescription => ({
+
+  const currentValue = get(childData, childLabelProp, '');
+
+  if (!childLabelProp) return currentValue;
```
actually this was correct and I readded it. I misunderstood the purpose of the line here.
